src/ORM/FieldType/DBBoolean.php

```diff
@@ -46,7 +46,7 @@ public function saveInto($dataObject)
     {
         $fieldName = $this->name;
         if ($fieldName) {
-            $dataObject->$fieldName = ($this->value) ? 1 : 0;
+            $dataObject->setFieldValue($fieldName, $this->value ? 1 : 0);
```

Comment:
Once you are inside `__set()` - in this case ViewableData::__set(), any futher call to `$dataObject->$fieldName` won't call `__set()` again as this would cause an infinite loop.  Instead at this point it's setting properties directly, which in most causes will be setting dynamic properties which are deprecated, hence the code change.

Let's say I'm call $MyDataObject->MyField = DBBoolean::create('MyField'); from PageController::init().  The code path is now:

```
ViewableData::__set() .. $this->setField($property, $value);
DataObject::setField() .. if ($val instanceof DBField) ..  $val->saveInto($this);
DBBoolean::saveInto() .. $dataObject->setFieldValue($fieldName, $this->value ? 1 : 0);
DataObject::setFieldValue() .. $this->record[$fieldName] = $val;
```

Reply:
Ahh okay, I was missing the context that it wouldn't call `__set()` again.

Can we not just call `setField()` here then instead of `setFieldValue()`? We don't need an entirely new public API to do this.

Reply:
No ... see the codepath above, at this point were are in DBBoolean::saveInto(), if we call DataObject::setField() we'll go into an infinite loop

Reply:
We won't go into an infinite loop, because the value that `DBBoolean::saveInto()` passes in is _not_ an instance of `DBField` - I suppose technically it could be, but that would be very unusual, and in any case it won't be _itself_ so it will eventually resolve down to a value.

Reply:
That would certainly be preferable to not have this new API, though I'm not sure that we can do this easily due to DBComposite.  There is unit testing to validate that you can actually set a dynamic property as a DBField currently. 
e.g. 
https://github.com/silverstripe/silverstripe-framework/blob/4/tests/php/ORM/DBCompositeTest.php#L33

There was this existing logic specifically for DBComposite in DataObject::setField()
```php
            // Situation 1a: Composite fields should remain bound in case they are
            // later referenced to update the parent dataobject
            if ($val instanceof DBComposite) {
                $val->bindTo($this);
                $this->record[$fieldName] = $val; // <<<<<< this
            }
```

Reply:
I don't understand the problem.... the test that you've pointed out passes if you call `setField()` instead of `setFieldValue()`. I can't see any problems with calling `setField()` where you're currently calling `setFieldValue()` in the various `DBField` classes.

Reply:
Oh right, because we're retaining the line I just highlighted :laughing: